framework/quota/src/org/apache/cloudstack/quota/dao/ServiceOfferingDaoImpl.java

```
@@ -0,0 +1,85 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+package org.apache.cloudstack.quota.dao;
+
+import java.util.Map;
+
+import javax.ejb.Local;
+import javax.inject.Inject;
+
+import org.apache.log4j.Logger;
+import org.springframework.stereotype.Component;
+import org.apache.cloudstack.quota.vo.ServiceOfferingVO;
+
+import com.cloud.event.UsageEventVO;
+import com.cloud.utils.db.DB;
+import com.cloud.utils.db.GenericDaoBase;
+import com.cloud.utils.db.TransactionLegacy;
+import com.cloud.utils.exception.CloudRuntimeException;
+
+@Component
+@Local(value = { ServiceOfferingDao.class })
+@DB()
+public class ServiceOfferingDaoImpl extends GenericDaoBase<ServiceOfferingVO, Long> implements ServiceOfferingDao {
+    protected static final Logger s_logger = Logger.getLogger(ServiceOfferingDaoImpl.class);
+
+    @Inject
+    UserVmDetailsDao userVmDetailsDao;
+
+    @Override
+    public ServiceOfferingVO findServiceOffering(final Long vmId, final long serviceOfferingId) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.CLOUD_DB);
+        ServiceOfferingVO result;
+        try {
+            result = findById(vmId, serviceOfferingId);
+        } finally {
+            txn.close();
+        }
+        TransactionLegacy.open(opendb).close();
+        return result;
+    }
+
+    public ServiceOfferingVO findById(Long vmId, long serviceOfferingId) {
+        ServiceOfferingVO offering = super.findById(serviceOfferingId);
+        if (offering.isDynamic()) {
+            offering.setDynamicFlag(true);
+            if (vmId == null) {
+                throw new CloudRuntimeException("missing argument vmId");
+            }
```

By convention, parameter checks should be at the top of the method to fail fast.  The ``com.google.common.base.Preconditions#checkArgument`` method provides a concise, syntactic sugar for checking arguments.